config.man.txt
/*
 * --------------------
 * Application settings
 * --------------------
 */

"app": {
	"logColors": true,
		(boolean)
		Use colored loglevels in the log-files. Loglevels will always appear colored in
		console, regardless of this setting.

	"ircDebugging": false,
		(boolean)
		Show raw output from the IRC-module in console. Raw messages from the IRC-module
		will never appear in logfiles, regardless of this setting.

	"cleanLogs": 10,
		(integer)
		How old, in days, a logfile gets to age before getting purged. Purge occurs
		every 24 hours, the first purge being 24 hours after application startup.

	"autoUpdate": "patch",
		(string)
		Allows Middleman to auto-update itself whenever it finds a new version through Git.
		Scans for new versions occur every 24 hours, and auto-updating will restart the app.

		Values:
			null				Turns off auto-update feature
			"patch" (default)	Auto-updates only on new patch releases (ex: 1.3.1 -> 1.3.2)
			"minor"				Auto-updates on new patch and minor releases (ex: 1.3.4 -> 1.4.2)
			"major"				Auto-updates on all new releases (ex: 1.3.6 -> 2.0.0)

	"hooks": ""
		(string)
		Custom script to run hooks after Middleman events. The event will be passed as
		first argument to the script, and the module will be passed as second argument.

		Examples:
			Bash:				"/bin/bash /path/to/hook.sh"
			Node.js:			"/usr/bin/node /path/to/hook.js"
			Python:				"/usr/bin/python /path/to/hook.py"
			Batch (Windows):	"C:\\Windows\\System32\\cmd.exe /C C:\\path\\to\\hook.bat"

		Events:
			connect			Connected to an IRC server
			disconnect		Disconnected from an IRC server
			download		Download-command sent to CouchPotato or SickBeard
			freeleech		Downloaded a freeleech-torrent

		Modules:
			btnirc
			ptpirc
			ptpanti
},


/*
 * ------------
 * PTP settings
 * ------------
 */

"ptp": {
	"username": "my_username",
		(string)
		Username on PTP

	"password": "my_password",
		(string)
		Password on PTP

	"passkey": "my_passkey",
		(string)
		Passkey on PTP

	"irckey": "my_irckey",
		(string)
		IRC-key on PTP


	/*
	 * Antitrump module settings
	 */
	"antitrump": {
		"enabled": false
			(boolean)
			Enables this module.

		"interval": 15,
			(integer)
			The interval, in minutes, to check for deleted snatches.

		"maxage": 86400,
			(integer)
			The maximum age, in seconds, of a torrent before it is considered too old.

		"maxtrumps": 3,
			(integer)
			The maximum amount a movie can be re-added before being ignored.

		"testmode": false
			(boolean)
			If enabled, no final commands will be sent to CouchPotato.
	},

	/*
	 * IRC-watch module settings
	 */
	"ircwatch": {
		"enabled": false,
			(boolean)
			Enables this module.

		"maxage": 1,
			(integer)
			Maximum age, in years, for an announce to be picked up and checked. If you
			only use CouchPotato for new movies, it's unnecessary to process announces
			for old releases. Does not apply to freeleech-torrents when freeleech-mode
			is enabled.

		"useIrcSSL": false,
			(boolean)
			Enables TLS and connects to the IRC server using the SSL port.

		"ignoreInvalidIrcSSL": false,
			(boolean)
			If the IRC server has an invalid SSL certificate when connecting securely,
			setting this options to true will ignore that and connect anyway.

		"nickname": "my_nick|AutoDL",
			(string)
			IRC nickname. Should end in "|AutoDL" and needs to be registered with
			NickServ or grouped with another NickServ account.

		"nickservPassword": "my_password",
			(string)
			NickServ password.

		"masterAuths": [""],
			(array)
			Whitelist of NickServ-auths who are allowed to execute commands on the bot.
			ex: ["myMaster"]					NickServ-auth "myMaster" is whitelisted.
			ex: ["myMaster", "otherMaster"]		Two NickServ-auths are whitelisted, etc.

		"masterTimeout": 600,
			(integer)
			Max inactivity, in seconds, before a master nick is unregistered with the bot.

		"freeleechDir": "/path/to/my/dir"
			(string)
			Absolute path to the destination where freeleech-torrents should be saved.
			ex: "/path/to/my/unix/style/dir"	UNIX-style path
			ex: "C:\\my\\windows\\style\\dir"	Windows-style path
	}
},


/*
 * ------------
 * BTN settings
 * ------------
 */

"btn": {
	"apikey": "my_apikey"
		(string)
		API-key on BTN

	/*
	 * IRC-watch module settings
	 */
	"ircwatch": {
		"enabled": false,
			(boolean)
			Enables this module.

		"allowForeign": false,
			(boolean)
			Should we notify SickBeard when there is a foreign release for an English show?

		"useIrcSSL": false,
			(boolean)
			Enables TLS and connects to the IRC server using the SSL port.

		"ignoreInvalidIrcSSL": false,
			(boolean)
			If the IRC server has an invalid SSL certificate when connecting securely,
			setting this options to true will ignore that and connect anyway.

		"nickname": "my_nick|AutoDL",
			(string)
			IRC nickname. Should end in "|AutoDL" and needs to be registered with
			NickServ or grouped with another NickServ account.

		"nickservPassword": "my_password",
			(string)
			NickServ password.

		"masterAuths": [""],
			(array)
			Whitelist of NickServ-auths who are allowed to execute commands on the bot.
			ex: ["myMaster"]					NickServ-auth "myMaster" is whitelisted.
			ex: ["myMaster", "otherMaster"]		Two NickServ-auths are whitelisted, etc.

		"masterTimeout": 600
			(integer)
			Max inactivity, in seconds, before a master nick is unregistered with the bot.
	}
},


/*
 * --------------------
 * CouchPotato settings
 * --------------------
 */

"cp": {
	"url": "http://<hostname>:<port>",
		(string)
		CouchPotato URL.

	"apikey": "my_apikey"
		(string)
		CouchPotato API-key.
},


/*
 * ------------------
 * SickBeard settings
 * ------------------
 */

"sb": {
	"ttl": 60,
		(integer)
		Time-To-Live for in-application SickBeard-cache.

	"url": "http://<sickbeard>:<port>",
		(string)
		SickBeard API-URL.

	"apikey": "my_apikey",
		(string)
		SickBeard API-key.

	"username": "my_username",
		(string)
		SickBeard username. Needed for running backlog-scans on shows, since it's not
		available from the API yet.

	"password": "my_password"
		(string)
		SickBeard password. Needed for running backlog-scans on shows, since it's not
		available from the API yet.
}


/*
 * -------------
 * HTTP settings
 * -------------
 */

"http": {
	"timeout": 30,
		(integer)
		HTTP request timeout in seconds

	"ignoreInvalidSSL": false
		(boolean)
		Ignores certificate warnings on HTTPS-requests. If you're running SickBeard or
		CouchPotato using a self-signed cert, you want this.
}


lib/util/config.js
var async = require('async');
var events = require('events');
var util = require('util');
var help = require('./helpers');

var db = require('./cache')('config', {exitOnError: true});

var defaults = {
	app: {
		logColors: true,
		ircDebugging: false,
		cleanLogs: 10,
		autoUpdate: 'patch',
		hooks: ''
	},
	ptp: {
		username: 'my_username',
		password: 'my_password',
		passkey: 'my_passkey',
		irckey: 'my_irckey',
		antitrump: {
			enabled: false,
			interval: 15,
			maxage: 86400,
			maxtrumps: 3,
			testmode: false
		},
		ircwatch: {
			enabled: false,
			maxage: 1,
			useIrcSSL: false,
			ignoreInvalidIrcSSL: false,
			nickname: 'my_nick|AutoDL',
			nickservPassword: 'my_password',
			masterAuths: [''],
			masterTimeout: 600,
			freeleechDir: '/path/to/my/dir'
		}
	},
	btn: {
		apikey: 'my_apikey',
		ircwatch: {
			enabled: false,
			allowForeign: false,
			useIrcSSL: false,
			ignoreInvalidIrcSSL: false,
			nickname: 'my_nick|AutoDL',
			nickservPassword: 'my_password',
			masterAuths: [''],
			masterTimeout: 600
		}
	},
	cp: {
		url: 'http://<hostname>:<port>',
		apikey: 'my_apikey'
	},
	sb: {
		ttl: 60,
		url: 'http://<sickbeard>:<port>',
		apikey: 'my_apikey',
		username: 'my_username',
		password: 'my_password'
	},
	http: {
		timeout: 30,
		ignoreInvalidSSL: false
	}
};

function mergeConfig(i, callback) {
	var v = db.get(i);
	if (!v) v = {};
	var c = d(defaults[i], v);
	db.set(i, c, callback);
}

function removeKey(k, callback) {
	db.remove(k, callback);
}

function cleanup(callback) {
	var dk = Object.keys(defaults);
	var ak = db.list();
	var series = [];

	for (var i in ak) {
		if (!help.inArray(ak[i], dk))
			series.push(async.apply(removeKey, ak[i]));
	}

	async.series(series, callback);
};

function d(defaults, config) {
	for (var i in defaults) {
		if (Object.prototype.toString.call(defaults[i]) == '[object Object]')
			defaults[i] = (typeof config[i] !== 'undefined') ? d(defaults[i], config[i]) : defaults[i];
		else defaults[i] = (typeof config[i] !== 'undefined') ? config[i] : defaults[i];
	}
	return defaults;
}

function Config() {
	events.EventEmitter.call(this);

	var self = this;
	var series = [];
	
	for (var i in defaults) series.push(async.apply(mergeConfig, i));
	series.push(cleanup);
	
	async.series(series, function(err) {
		if (err) console.log('ERROR: %s', err);

		var list = db.list();
		for (var i in list) {
			self.__dg(list[i]);
		}

		self.emit('ready');
	});
}
util.inherits(Config, events.EventEmitter);

Config.prototype.__dg = function(i) {
	this.__defineGetter__(i, function() {
		return db.get(i);
	});
};

module.exports = new Config();
